compiler/rustc_middle/src/ty/instance.rs

```diff
@@ -539,8 +539,17 @@ impl<'tcx> Instance<'tcx> {
     }
 
     pub fn resolve_drop_in_place(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> ty::Instance<'tcx> {
+        let ty_erased =
+            if let (true, TyKind::FnPtr(poly_fn_sig)) = (ty.has_late_bound_regions(), ty.kind()) {
+                let re_erased = tcx.erase_late_bound_regions(*poly_fn_sig);
+                let rebound = Binder::dummy(re_erased);
+                let reassembled = tcx.mk_fn_ptr(rebound);
+                reassembled
+            } else {
+                ty
+            };
```

Comment:
```suggestion
        // Erase regions before calling resolve, as they must not change the resolution.
        let ty_erased =tcx.erase_late_bound_regions(ty);
```

Reply:
I think I tried that but normalizing late bound regions is not enough in this case. There's still a mismatch because one symbol looks like `core::ptr::drop_in_place::<for<'a> fn(&'a ())>` and the other like `core::ptr::drop_in_place::<fn(&())>`.

Reply:
Changed my suggestion to use `erase_regions` that should erase everything.

Reply:
Well, not sure actually. It's probably the instance resolution code that needs to be fortified against leaking regions.

Reply:
> Changed my suggestion to use `erase_regions` that should erase everything.

I believe this already happening without desired effect wrt to this bug: `Instance::expect_resolve` -> `Instance::resolve` -> `Instance::resolve_opt_const_arg` -> `tcx.erase_regions(substs)`. Anyway, I'm going to give it a try.

Reply:
> everything.

That's what I thought too, but it doesn't seem to be the case:
> /// Returns an equivalent value with all free regions removed (note
> /// that late-bound regions remain, because they are important for
> /// subtyping, but they are anonymized and normalized as well)..
> pub fn erase_regions<T>(self, value: T) -> T

Reply:
Just checked, it doesn't erase late bound regions.

Reply:
https://github.com/rust-lang/rust/blob/b5c8c329a7b4f6a12ae72ae41cb834989228221e/compiler/rustc_middle/src/ty/erase_regions.rs#L64
this is the `TypeFolder` invoked during `erase_regions`